File: Lib/Telerik/Examples/DemoAppsHub/ListViewExample/ListViewExampleUserControl.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using Telerik.Examples.WinControls.DataSources;
using System.IO;
using Telerik.WinControls.UI;
using Telerik.WinControls.Enumerations;
using Telerik.WinControls.Data;
using Telerik.WinControls.Primitives;
using Telerik.WinControls.Layouts;

namespace DemoAppsHub.ListViewExample
{
    public partial class ListViewExampleUserControl : UserControl
    {
        List<string> features = new List<string>() { "AirConditioner", "Mp3Player", "DVDPlayer", "ABS", "ASR", "Navigation" };
        private bool updatingToggleState = false;

        public ListViewExampleUserControl()
        {
            InitializeComponent();

            this.radCommandBar1.ThemeName = "TelerikMetroBlue";
            this.radListView1.ThemeName = "TelerikMetroBlue";

            ImagePrimitive searchIcon = new ImagePrimitive();
            searchIcon.Image = DemoAppsHub.Properties.Resources.TV_search;
            searchIcon.Alignment = ContentAlignment.MiddleRight;
            
            this.commandBarTextBoxFilter.TextBoxElement.StretchHorizontally = true;
            this.commandBarTextBoxFilter.TextBoxElement.Alignment = ContentAlignment.MiddleRight;
            this.commandBarTextBoxFilter.TextBoxElement.Children.Add(searchIcon);
            this.commandBarTextBoxFilter.TextBoxElement.TextBoxItem.Alignment = ContentAlignment.MiddleLeft;
            this.commandBarTextBoxFilter.TextBoxElement.TextBoxItem.PropertyChanged += new PropertyChangedEventHandler(TextBoxItem_PropertyChanged);

            this.radListView1.VisualItemFormatting += new Telerik.WinControls.UI.ListViewVisualItemEventHandler(radListView1_VisualItemFormatting);
            this.radListView1.ViewTypeChanged += new EventHandler(radListView1_ViewTypeChanged);
            this.radListView1.CellFormatting += new Telerik.WinControls.UI.ListViewCellFormattingEventHandler(radListView1_CellFormatting);

            this.radListView1.AllowEdit = false;
            this.radListView1.AllowRemove = false;

            this.radListView1_ViewTypeChanged(this, EventArgs.Empty);

            this.commandBarToggleDetails.MinSize = new Size(35, 35);
            this.commandBarToggleList.MinSize = new Size(35, 35);
            this.commandBarToggleTiles.MinSize = new Size(35, 35);

            this.commandBarTextBoxFilter.StretchVertically = true;
            this.commandBarTextBoxFilter.TextBoxElement.StretchVertically = true;

            this.commandBarDropDownGroup.StretchVertically = true;
            this.commandBarDropDownGroup.DropDownListElement.ItemHeight = 35;
            this.commandBarDropDownSort.StretchVertically = true;
            this.commandBarDropDownSort.DropDownListElement.ItemHeight = 35;
            this.commandBarStripElement1.OverflowButton.MinSize = new Size(35, 0);

        }

        private void SetupDetailsView()
        {
            this.radListView1.ItemSize = new Size(0, 110);
        }

        private void SetupIconsView()
        {
            this.radListView1.ItemSize = new Size(295, 120);
            this.radListView1.ItemSpacing = 5;
            this.radListView1.GroupIndent = 0;
        }

        private void SetupSimpleListView()
        {
            this.radListView1.AllowArbitraryItemHeight = true;

        }

        private bool ContainsFeature(ListViewDataItem item, string feature)
        {
            return item[feature] != null && Convert.ToInt32(item[feature]) != 0;
        }

        private string GetFeatures(ListViewDataItem item)
        {
            StringBuilder featuresString = new StringBuilder();

            foreach (string feature in this.features)
            {
                if (ContainsFeature(item, feature))
                {
                    featuresString.Append(feature + ", ");
                }
            }

            if (featuresString.Length > 1)
            {
                featuresString.Remove(featuresString.Length - 2, 2);
            }

            return featuresString.ToString();
        }
        
        #region Event handlers

        protected override void OnLoad(EventArgs e)
        {
            base.OnLoad(e);
            this.carsRatesDataTableTableAdapter.Fill(this.sofiaCarRentalDataSet.CarsRatesDataTable);

            this.commandBarDropDownGroup.SelectedIndex = 1;
        }

        void radListView1_BindingCompleted(object sender, System.EventArgs e)
        {
            this.radListView1.Columns["ImageFileName"].Width = 180;
            this.radListView1.Columns["ImageFileName"].MinWidth = 180;
            this.radListView1.Columns["Make"].Width = 90;
            this.radListView1.Columns["Make"].MinWidth = 90;
            this.radListView1.Columns["Model"].Width = 110;
            this.radListView1.Columns["Model"].MinWidth = 110;
            this.radListView1.Columns["CarYear"].Width = 90;
            this.radListView1.Columns["CarYear"].MinWidth = 90;
            this.radListView1.Columns["CategoryName"].Width = 90;
            this.radListView1.Columns["CategoryName"].MinWidth = 90;

            int pictureColumnIndex = this.radListView1.Columns.IndexOf("ImageFileName");
            this.radListView1.Columns.Move(pictureColumnIndex, 0);
        }

        void radListView1_ColumnCreating(object sender, ListViewColumnCreatingEventArgs e)
        {
            if (e.Column.FieldName == "CarID")
                e.Column.Visible = false;

            if (e.Column.FieldName == "ImageFileName")
                e.Column.HeaderText = "Picture";

            if (e.Column.FieldName == "CarYear")
                e.Column.HeaderText = "Manufactured";

            if (e.Column.FieldName == "CategoryName")
                e.Column.HeaderText = "Category";

            if (e.Column.FieldName == "Mp3Player")
                e.Column.HeaderText = "MP3";

            if (e.Column.FieldName == "DVDPlayer")
                e.Column.HeaderText = "DVD";

            if (e.Column.FieldName == "AirConditioner")
            {
                e.Column.HeaderText = "Air Cond.";
                e.Column.Width = 90;
                e.Column.MinWidth = 90;
            }

            if (e.Column.FieldName == "Daily" || e.Column.FieldName == "Weekly" || e.Column.FieldName == "Monthly" || e.Column.FieldName == "Available")
                e.Column.Visible = false;

            if (features.Contains(e.Column.FieldName))
            {
                e.Column.Width = 55;
                e.Column.MinWidth = 55;
            }
        }

        void radListView1_CellFormatting(object sender, ListViewCellFormattingEventArgs e)
        {

            if (e.CellElement is DetailListViewHeaderCellElement)
            {
                return;
            }

            if (e.CellElement.Data.HeaderText == "Picture")
            {
                ((DetailListViewDataCellElement)e.CellElement).Image = ((DetailListViewDataCellElement)e.CellElement).Row.Image;
                e.CellElement.Text = "";
                e.CellElement.ImageAlignment = ContentAlignment.MiddleCenter;
                e.CellElement.TextImageRelation = TextImageRelation.Overlay;
            }
            else
            {
                e.CellElement.Image = null;
            }

            if (e.CellElement.Data.HeaderText == "Make" || e.CellElement.Data.HeaderText == "Model")
            {
                e.CellElement.Text = "<html><span style=\"color:#161112;font-size:11.5pt;\">" + e.CellElement.Text + "</span>";
            }
            else if (this.features.Contains(e.CellElement.Data.FieldName))
            {
                bool containsFeature = this.ContainsFeature(((DetailListViewDataCellElement)e.CellElement).Row, e.CellElement.Data.FieldName);
                string color = (containsFeature) ? "#050F70" : "#B52822";
                e.CellElement.ForeColor = (Color)(new ColorConverter().ConvertFromString(color));
                e.CellElement.Font = new Font(e.CellElement.Font.FontFamily, 10, GraphicsUnit.Point);
                e.CellElement.Text = (containsFeature) ? "YES" : "NO";
            }
            else if (e.CellElement.Data.HeaderText != "Picture")
            {
                e.CellElement.ForeColor = (Color)(new ColorConverter().ConvertFromString("#050F70"));
                e.CellElement.Font = new Font(e.CellElement.Font.FontFamily, 10, GraphicsUnit.Point);
                e.CellElement.Text = e.CellElement.Text;
            }
        }

        void radListView1_ViewTypeChanged(object sender, EventArgs e)
        {
            switch (radListView1.ViewType)
            {
                case ListViewType.ListView:
                    SetupSimpleListView();
                    break;
                case ListViewType.IconsView:
                    SetupIconsView();
                    break;
                case ListViewType.DetailsView:
                    SetupDetailsView();
                    break;
            }
        }

        void radListView1_VisualItemCreating(object sender, Telerik.WinControls.UI.ListViewVisualItemCreatingEventArgs e)
        {
            if (this.radListView1.ViewType == ListViewType.ListView && !(e.VisualItem is BaseListViewGroupVisualItem))
            {
                e.VisualItem = new CarsListVisualItem();
            }
        }

        void radListView1_VisualItemFormatting(object sender, Telerik.WinControls.UI.ListViewVisualItemEventArgs e)
        {
            if (e.VisualItem is BaseListViewGroupVisualItem)
            {
                return;
            }

            if (this.radListView1.ViewType == ListViewType.IconsView)
            {
                e.VisualItem.Font = new System.Drawing.Font("Segoe UI", 12);
                e.VisualItem.ForeColor = Color.Black;
                e.VisualItem.Text = "<html>" +
                   "<span style=\"color:#040203;font-size:12pt;\">" + e.VisualItem.Data["Make"] + " " + e.VisualItem.Data["Model"] + "</span>" +
                   "<br><span style=\"color:#040203;font-size:9pt;\">" + e.VisualItem.Data["CarYear"] + ", " + e.VisualItem.Data["CategoryName"] + "</span>" +
                   "<br><br><span style=\"color:#112164;font-size:9pt;\">" + GetFeatures(e.VisualItem.Data) + "</span>";

                e.VisualItem.ImageLayout = ImageLayout.Center;
                e.VisualItem.ImageAlignment = ContentAlignment.MiddleCenter;
            } 

            if (this.radListView1.ViewType == ListViewType.ListView)
            {
                e.VisualItem.Padding = new Padding(5, 5, 0, 5);
                e.VisualItem.Layout.LeftPart.Margin = new Padding(0, 0, 5, 0);
            }
        }

        void radListView1_ItemDataBound(object sender, Telerik.WinControls.UI.ListViewItemEventArgs e)
        {
            e.Item.Image = Image.FromFile(Application.StartupPath + @"\CarRentalImages\" + e.Item["ImageFileName"]);
        }

        void viewToggleButton_ToggleStateChanged(object sender, StateChangedEventArgs args)
        {
            if (updatingToggleState)
            {
                return;
            }

            this.updatingToggleState = true;

            if (this.commandBarToggleDetails != sender)
            {
                this.commandBarToggleDetails.ToggleState = ToggleState.Off;
            }

            if (this.commandBarToggleList != sender)
            {
                this.commandBarToggleList.ToggleState = ToggleState.Off;
            }

            if (this.commandBarToggleTiles != sender)
            {
                this.commandBarToggleTiles.ToggleState = ToggleState.Off;
            }

            this.updatingToggleState = false;

            if (this.commandBarToggleDetails.ToggleState == ToggleState.On)
            {
                this.radListView1.ViewType = ListViewType.DetailsView;
            }

            if (this.commandBarToggleList.ToggleState == ToggleState.On)
            {
                this.radListView1.ViewType = ListViewType.ListView;
            }

            if (this.commandBarToggleTiles.ToggleState == ToggleState.On)
            {
                this.radListView1.ViewType = ListViewType.IconsView;
            }

        }

        void viewToggleButton_ToggleStateChanging(object sender, StateChangingEventArgs args)
        {
            if (!updatingToggleState && args.OldValue == ToggleState.On)
            {
                args.Cancel = true;
            }
        }

        void commandBarDropDownSort_SelectedIndexChanged(object sender, Telerik.WinControls.UI.Data.PositionChangedEventArgs e)
        {
            this.radListView1.SortDescriptors.Clear();
            switch (this.commandBarDropDownSort.Text)
            {
                case "Make":
                    this.radListView1.SortDescriptors.Add(new SortDescriptor("Make", ListSortDirection.Ascending));
                    this.radListView1.EnableSorting = true;
                    break;
                case "Model":
                    this.radListView1.SortDescriptors.Add(new SortDescriptor("Model", ListSortDirection.Ascending));
                    this.radListView1.EnableSorting = true;
                    break;
                case "Category":
                    this.radListView1.SortDescriptors.Add(new SortDescriptor("CategoryName", ListSortDirection.Ascending));
                    this.radListView1.EnableSorting = true;
                    break;
                case "Year":
                    this.radListView1.SortDescriptors.Add(new SortDescriptor("CarYear", ListSortDirection.Ascending));
                    this.radListView1.EnableSorting = true;
                    break;
            }
        }
        
        void commandBarDropDownGroup_SelectedIndexChanged(object sender, Telerik.WinControls.UI.Data.PositionChangedEventArgs e)
        {
            this.radListView1.GroupDescriptors.Clear();
            switch (this.commandBarDropDownGroup.Text)
            {
                case "None":
                    this.radListView1.EnableGrouping = false;
                    this.radListView1.ShowGroups = false;
                    break;
                case "Make":
                    this.radListView1.GroupDescriptors.Add(new GroupDescriptor(
                        new SortDescriptor[] { new SortDescriptor("Make", ListSortDirection.Ascending) }));
                    this.radListView1.EnableGrouping = true;
                    this.radListView1.ShowGroups = true;
                    break;
                case "Category":
                    this.radListView1.GroupDescriptors.Add(new GroupDescriptor(
                        new SortDescriptor[] { new SortDescriptor("CategoryName", ListSortDirection.Ascending) }));
                    this.radListView1.EnableGrouping = true;
                    this.radListView1.ShowGroups = true;
                    break;
                case "Year":
                    this.radListView1.GroupDescriptors.Add(new GroupDescriptor(
                        new SortDescriptor[] { new SortDescriptor("CarYear", ListSortDirection.Ascending) }));
                    this.radListView1.EnableGrouping = true;
                    this.radListView1.ShowGroups = true;
                    break;
            }
        }

        void commandBarTextBoxFilter_TextChanged(object sender, EventArgs e)
        {
            this.radListView1.FilterDescriptors.Clear();

            if (String.IsNullOrEmpty(this.commandBarTextBoxFilter.Text))
            {
                this.radListView1.EnableFiltering = false;
            }
            else
            {
                this.radListView1.FilterDescriptors.LogicalOperator = FilterLogicalOperator.Or;
                this.radListView1.FilterDescriptors.Add("Make", FilterOperator.Contains, this.commandBarTextBoxFilter.Text);
                this.radListView1.FilterDescriptors.Add("Model", FilterOperator.Contains, this.commandBarTextBoxFilter.Text);
                this.radListView1.EnableFiltering = true;
            }
        }

        void TextBoxItem_PropertyChanged(object sender, PropertyChangedEventArgs e)
        {
            if (e.PropertyName == "Bounds")
            {
                commandBarTextBoxFilter.TextBoxElement.TextBoxItem.HostedControl.MaximumSize = new Size((int)commandBarTextBoxFilter.DesiredSize.Width - 28, 0);
            }
        }

        #endregion
    }
}

File: Lib/Telerik/Examples/DemoAppsHub/ListViewExample/ListViewExampleUserControl.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports System.IO
Imports Telerik.WinControls.UI
Imports Telerik.WinControls.Enumerations
Imports Telerik.WinControls.Data
Imports Telerik.WinControls.Primitives
Imports Telerik.WinControls.Layouts

Namespace DemoAppsHub.ListViewExample
	Public Partial Class ListViewExampleUserControl
		Inherits UserControl
        Private features As List(Of String) = New List(Of String)() From {"AirConditioner", "Mp3Player", "DVDPlayer", "ABS", "ASR", "Navigation"}
		Private updatingToggleState As Boolean = False

		Public Sub New()
			InitializeComponent()

			Me.radCommandBar1.ThemeName = "TelerikMetroBlue"
			Me.radListView1.ThemeName = "TelerikMetroBlue"

			Dim searchIcon As ImagePrimitive = New ImagePrimitive()
			searchIcon.Image = Global.Resources.TV_search
			searchIcon.Alignment = ContentAlignment.MiddleRight

			Me.commandBarTextBoxFilter.TextBoxElement.StretchHorizontally = True
			Me.commandBarTextBoxFilter.TextBoxElement.Alignment = ContentAlignment.MiddleRight
			Me.commandBarTextBoxFilter.TextBoxElement.Children.Add(searchIcon)
			Me.commandBarTextBoxFilter.TextBoxElement.TextBoxItem.Alignment = ContentAlignment.MiddleLeft
			AddHandler commandBarTextBoxFilter.TextBoxElement.TextBoxItem.PropertyChanged, AddressOf TextBoxItem_PropertyChanged

			AddHandler radListView1.VisualItemFormatting, AddressOf radListView1_VisualItemFormatting
			AddHandler radListView1.ViewTypeChanged, AddressOf radListView1_ViewTypeChanged
			AddHandler radListView1.CellFormatting, AddressOf radListView1_CellFormatting

			Me.radListView1.AllowEdit = False
			Me.radListView1.AllowRemove = False

			Me.radListView1_ViewTypeChanged(Me, EventArgs.Empty)

			Me.commandBarToggleDetails.MinSize = New Size(35, 35)
			Me.commandBarToggleList.MinSize = New Size(35, 35)
			Me.commandBarToggleTiles.MinSize = New Size(35, 35)

			Me.commandBarTextBoxFilter.StretchVertically = True
			Me.commandBarTextBoxFilter.TextBoxElement.StretchVertically = True

			Me.commandBarDropDownGroup.StretchVertically = True
			Me.commandBarDropDownGroup.DropDownListElement.ItemHeight = 35
			Me.commandBarDropDownSort.StretchVertically = True
			Me.commandBarDropDownSort.DropDownListElement.ItemHeight = 35
			Me.commandBarStripElement1.OverflowButton.MinSize = New Size(35, 0)

		End Sub

		Private Sub SetupDetailsView()
			Me.radListView1.ItemSize = New Size(0, 110)
		End Sub

		Private Sub SetupIconsView()
			Me.radListView1.ItemSize = New Size(295, 120)
			Me.radListView1.ItemSpacing = 5
			Me.radListView1.GroupIndent = 0
		End Sub

		Private Sub SetupSimpleListView()
			Me.radListView1.AllowArbitraryItemHeight = True

		End Sub

		Private Function ContainsFeature(ByVal item As ListViewDataItem, ByVal feature As String) As Boolean
			Return Not item(feature) Is Nothing AndAlso Convert.ToInt32(item(feature)) <> 0
		End Function

		Private Function GetFeatures(ByVal item As ListViewDataItem) As String
			Dim featuresString As StringBuilder = New StringBuilder()

			For Each feature As String In Me.features
				If ContainsFeature(item, feature) Then
					featuresString.Append(feature & ", ")
				End If
			Next feature

			If featuresString.Length > 1 Then
				featuresString.Remove(featuresString.Length - 2, 2)
			End If

			Return featuresString.ToString()
		End Function

		#Region "Event handlers"

		Protected Overrides Sub OnLoad(ByVal e As EventArgs)
			MyBase.OnLoad(e)
			Me.carsRatesDataTableTableAdapter.Fill(Me.sofiaCarRentalDataSet.CarsRatesDataTable)

			Me.commandBarDropDownGroup.SelectedIndex = 1
		End Sub

		Private Sub radListView1_BindingCompleted(ByVal sender As Object, ByVal e As System.EventArgs) Handles radListView1.BindingCompleted
			Me.radListView1.Columns("ImageFileName").Width = 180
			Me.radListView1.Columns("ImageFileName").MinWidth = 180
			Me.radListView1.Columns("Make").Width = 90
			Me.radListView1.Columns("Make").MinWidth = 90
			Me.radListView1.Columns("Model").Width = 110
			Me.radListView1.Columns("Model").MinWidth = 110
			Me.radListView1.Columns("CarYear").Width = 90
			Me.radListView1.Columns("CarYear").MinWidth = 90
			Me.radListView1.Columns("CategoryName").Width = 90
			Me.radListView1.Columns("CategoryName").MinWidth = 90

			Dim pictureColumnIndex As Integer = Me.radListView1.Columns.IndexOf("ImageFileName")
			Me.radListView1.Columns.Move(pictureColumnIndex, 0)
		End Sub

		Private Sub radListView1_ColumnCreating(ByVal sender As Object, ByVal e As ListViewColumnCreatingEventArgs) Handles radListView1.ColumnCreating
			If e.Column.FieldName = "CarID" Then
				e.Column.Visible = False
			End If

			If e.Column.FieldName = "ImageFileName" Then
				e.Column.HeaderText = "Picture"
			End If

			If e.Column.FieldName = "CarYear" Then
				e.Column.HeaderText = "Manufactured"
			End If

			If e.Column.FieldName = "CategoryName" Then
				e.Column.HeaderText = "Category"
			End If

			If e.Column.FieldName = "Mp3Player" Then
				e.Column.HeaderText = "MP3"
			End If

			If e.Column.FieldName = "DVDPlayer" Then
				e.Column.HeaderText = "DVD"
			End If

			If e.Column.FieldName = "AirConditioner" Then
				e.Column.HeaderText = "Air Cond."
				e.Column.Width = 90
				e.Column.MinWidth = 90
			End If

			If e.Column.FieldName = "Daily" OrElse e.Column.FieldName = "Weekly" OrElse e.Column.FieldName = "Monthly" OrElse e.Column.FieldName = "Available" Then
				e.Column.Visible = False
			End If

			If features.Contains(e.Column.FieldName) Then
				e.Column.Width = 55
				e.Column.MinWidth = 55
			End If
		End Sub

		Private Sub radListView1_CellFormatting(ByVal sender As Object, ByVal e As ListViewCellFormattingEventArgs)

			If TypeOf e.CellElement Is DetailListViewHeaderCellElement Then
				Return
			End If

			If e.CellElement.Data.HeaderText = "Picture" Then
				CType(e.CellElement, DetailListViewDataCellElement).Image = (CType(e.CellElement, DetailListViewDataCellElement)).Row.Image
				e.CellElement.Text = ""
				e.CellElement.ImageAlignment = ContentAlignment.MiddleCenter
				e.CellElement.TextImageRelation = TextImageRelation.Overlay
			Else
				e.CellElement.Image = Nothing
			End If

			If e.CellElement.Data.HeaderText = "Make" OrElse e.CellElement.Data.HeaderText = "Model" Then
				e.CellElement.Text = "<html><span style=""color:#161112;font-size:11.5pt;"">" & e.CellElement.Text & "</span>"
			ElseIf Me.features.Contains(e.CellElement.Data.FieldName) Then
				Dim containsFeature As Boolean = Me.ContainsFeature((CType(e.CellElement, DetailListViewDataCellElement)).Row, e.CellElement.Data.FieldName)
				Dim color As String
				If (containsFeature) Then
					color = "#050F70"
				Else
					color = "#B52822"
				End If
				e.CellElement.ForeColor = CType(New ColorConverter().ConvertFromString(color), Color)
				e.CellElement.Font = New Font(e.CellElement.Font.FontFamily, 10, GraphicsUnit.Point)
				If (containsFeature) Then
					e.CellElement.Text = "YES"
				Else
					e.CellElement.Text = "NO"
				End If
			ElseIf e.CellElement.Data.HeaderText <> "Picture" Then
				e.CellElement.ForeColor = CType(New ColorConverter().ConvertFromString("#050F70"), Color)
				e.CellElement.Font = New Font(e.CellElement.Font.FontFamily, 10, GraphicsUnit.Point)
				e.CellElement.Text = e.CellElement.Text
			End If
		End Sub

		Private Sub radListView1_ViewTypeChanged(ByVal sender As Object, ByVal e As EventArgs)
			Select Case radListView1.ViewType
				Case ListViewType.ListView
					SetupSimpleListView()
				Case ListViewType.IconsView
					SetupIconsView()
				Case ListViewType.DetailsView
					SetupDetailsView()
			End Select
		End Sub

		Private Sub radListView1_VisualItemCreating(ByVal sender As Object, ByVal e As Telerik.WinControls.UI.ListViewVisualItemCreatingEventArgs) Handles radListView1.VisualItemCreating
			If Me.radListView1.ViewType = ListViewType.ListView AndAlso Not(TypeOf e.VisualItem Is BaseListViewGroupVisualItem) Then
				e.VisualItem = New CarsListVisualItem()
			End If
		End Sub

		Private Sub radListView1_VisualItemFormatting(ByVal sender As Object, ByVal e As Telerik.WinControls.UI.ListViewVisualItemEventArgs)
			If TypeOf e.VisualItem Is BaseListViewGroupVisualItem Then
				Return
			End If

			If Me.radListView1.ViewType = ListViewType.IconsView Then
				e.VisualItem.Font = New System.Drawing.Font("Segoe UI", 12)
				e.VisualItem.ForeColor = Color.Black
                e.VisualItem.Text = "<html>" & "<span style=""color:#040203;font-size:12pt;"">" & String.Format("{0}", e.VisualItem.Data("Make")) & _
                    " " & String.Format("{0}", e.VisualItem.Data("Model")) & "</span>" & "<br><span style=""color:#040203;font-size:9pt;"">" & _
                    String.Format("{0}", e.VisualItem.Data("CarYear")) & ", " & String.Format("{0}", e.VisualItem.Data("CategoryName")) & "</span>" & _
                    "<br><br><span style=""color:#112164;font-size:9pt;"">" & GetFeatures(e.VisualItem.Data) & "</span>"

				e.VisualItem.ImageLayout = ImageLayout.Center
				e.VisualItem.ImageAlignment = ContentAlignment.MiddleCenter
			End If

			If Me.radListView1.ViewType = ListViewType.ListView Then
				e.VisualItem.Padding = New Padding(5, 5, 0, 5)
				e.VisualItem.Layout.LeftPart.Margin = New Padding(0, 0, 5, 0)
			End If
		End Sub

		Private Sub radListView1_ItemDataBound(ByVal sender As Object, ByVal e As Telerik.WinControls.UI.ListViewItemEventArgs) Handles radListView1.ItemDataBound
            e.Item.Image = Image.FromFile(Application.StartupPath & "\CarRentalImages\" & String.Format("{0}", e.Item("ImageFileName")))
		End Sub

		Private Sub viewToggleButton_ToggleStateChanged(ByVal sender As Object, ByVal args As StateChangedEventArgs) Handles commandBarToggleList.ToggleStateChanged, commandBarToggleTiles.ToggleStateChanged, commandBarToggleDetails.ToggleStateChanged
			If updatingToggleState Then
				Return
			End If

			Me.updatingToggleState = True

			If Not Me.commandBarToggleDetails Is sender Then
				Me.commandBarToggleDetails.ToggleState = ToggleState.Off
			End If

			If Not Me.commandBarToggleList Is sender Then
				Me.commandBarToggleList.ToggleState = ToggleState.Off
			End If

			If Not Me.commandBarToggleTiles Is sender Then
				Me.commandBarToggleTiles.ToggleState = ToggleState.Off
			End If

			Me.updatingToggleState = False

			If Me.commandBarToggleDetails.ToggleState = ToggleState.On Then
				Me.radListView1.ViewType = ListViewType.DetailsView
			End If

			If Me.commandBarToggleList.ToggleState = ToggleState.On Then
				Me.radListView1.ViewType = ListViewType.ListView
			End If

			If Me.commandBarToggleTiles.ToggleState = ToggleState.On Then
				Me.radListView1.ViewType = ListViewType.IconsView
			End If

		End Sub

		Private Sub viewToggleButton_ToggleStateChanging(ByVal sender As Object, ByVal args As StateChangingEventArgs) Handles commandBarToggleList.ToggleStateChanging, commandBarToggleTiles.ToggleStateChanging, commandBarToggleDetails.ToggleStateChanging
			If (Not updatingToggleState) AndAlso args.OldValue = ToggleState.On Then
				args.Cancel = True
			End If
		End Sub

		Private Sub commandBarDropDownSort_SelectedIndexChanged(ByVal sender As Object, ByVal e As Telerik.WinControls.UI.Data.PositionChangedEventArgs) Handles commandBarDropDownSort.SelectedIndexChanged
			Me.radListView1.SortDescriptors.Clear()
			Select Case Me.commandBarDropDownSort.Text
				Case "Make"
					Me.radListView1.SortDescriptors.Add(New SortDescriptor("Make", ListSortDirection.Ascending))
					Me.radListView1.EnableSorting = True
				Case "Model"
					Me.radListView1.SortDescriptors.Add(New SortDescriptor("Model", ListSortDirection.Ascending))
					Me.radListView1.EnableSorting = True
				Case "Category"
					Me.radListView1.SortDescriptors.Add(New SortDescriptor("CategoryName", ListSortDirection.Ascending))
					Me.radListView1.EnableSorting = True
				Case "Year"
					Me.radListView1.SortDescriptors.Add(New SortDescriptor("CarYear", ListSortDirection.Ascending))
					Me.radListView1.EnableSorting = True
			End Select
		End Sub

		Private Sub commandBarDropDownGroup_SelectedIndexChanged(ByVal sender As Object, ByVal e As Telerik.WinControls.UI.Data.PositionChangedEventArgs) Handles commandBarDropDownGroup.SelectedIndexChanged
			Me.radListView1.GroupDescriptors.Clear()
			Select Case Me.commandBarDropDownGroup.Text
				Case "None"
					Me.radListView1.EnableGrouping = False
					Me.radListView1.ShowGroups = False
				Case "Make"
					Me.radListView1.GroupDescriptors.Add(New GroupDescriptor(New SortDescriptor() { New SortDescriptor("Make", ListSortDirection.Ascending) }))
					Me.radListView1.EnableGrouping = True
					Me.radListView1.ShowGroups = True
				Case "Category"
					Me.radListView1.GroupDescriptors.Add(New GroupDescriptor(New SortDescriptor() { New SortDescriptor("CategoryName", ListSortDirection.Ascending) }))
					Me.radListView1.EnableGrouping = True
					Me.radListView1.ShowGroups = True
				Case "Year"
					Me.radListView1.GroupDescriptors.Add(New GroupDescriptor(New SortDescriptor() { New SortDescriptor("CarYear", ListSortDirection.Ascending) }))
					Me.radListView1.EnableGrouping = True
					Me.radListView1.ShowGroups = True
			End Select
		End Sub

		Private Sub commandBarTextBoxFilter_TextChanged(ByVal sender As Object, ByVal e As EventArgs) Handles commandBarTextBoxFilter.TextChanged
			Me.radListView1.FilterDescriptors.Clear()

			If String.IsNullOrEmpty(Me.commandBarTextBoxFilter.Text) Then
				Me.radListView1.EnableFiltering = False
			Else
				Me.radListView1.FilterDescriptors.LogicalOperator = FilterLogicalOperator.Or
				Me.radListView1.FilterDescriptors.Add("Make", FilterOperator.Contains, Me.commandBarTextBoxFilter.Text)
				Me.radListView1.FilterDescriptors.Add("Model", FilterOperator.Contains, Me.commandBarTextBoxFilter.Text)
				Me.radListView1.EnableFiltering = True
			End If
		End Sub

		Private Sub TextBoxItem_PropertyChanged(ByVal sender As Object, ByVal e As PropertyChangedEventArgs)
			If e.PropertyName = "Bounds" Then
				commandBarTextBoxFilter.TextBoxElement.TextBoxItem.HostedControl.MaximumSize = New Size(CInt(Fix(commandBarTextBoxFilter.DesiredSize.Width - 28)), 0)
			End If
		End Sub

		#End Region
	End Class
End Namespace
